rfc/002-fault-tolerance/README.md

```diff
@@ -234,62 +240,58 @@ adds the requirement for a quick resynchronization ability to the metastore.
 emulate a reliable medium. For this proposal, working around an unreliable data
 transport channel is out of scope.
 
-Upon insertion of new data the catalog needs to be able to recognise whether
-the same data has already been ingested and discard the respective duplicate
-partitions.
-
-A possible detection mechanism can be achieved with per-source checksums on the
-written data. This requires a first architectural change: A partition may no
-longer contain data from multiple sources. Since it is the unit of work in the
-catalog, a partition can either be accepted or rejected in full.
-
-In addition to the checksums, partitions need to store source stream offsets as
-part of its metadata. An atomic commit of both partition UUID and its source
-position allows the source to pick up exactly at the point of failure, without
-risk of re-ingesting twice or missing any events. It thus provides exactly-once
-semantics for our write path.
-
-In technical terms, we can extend the CAF Streaming message type to include a
-handle to the source and source specific position, and then trigger an ACK once
-the data has been persisted. VAST already implements a similar mechanism today
-that acknowledges when events arrived at an active partition, which (almost)
-guarantees that they are queryable.
-
-An existing source can re-synchronize its position with the catalog to recover
-lost ACKs and advance its position to the location of data that was last
-commited. However, this is purely an optimization and not required for
-correctness.
+The unit of transaction with the catalog is a single partition. It is up to the
+source-specific data acquisition implementation to handle ACK'ing upstream data.
+For example, the Kafka source would forward its offset after having received an
+ACK from metastore about the corresponding partition.
+
+#### Duplicate Detection
+
+Detecting duplicates is frequent concern of users. We consider this topic as
+*client-side* functionality that is orthogonal to the atomic partition
+submission at the catalog. For example, a client may add a mechanism to keep
+track of already ingested files, e.g., by file name and/or hash digest.
+
+A specific scenario to consider during regular ingestion is the event that the
+server (and thereby the catalog) crashes during partition submission. In this
+case the ACK of the transaction may not arrive at the client. The client would
+then have to re-submit the same partition after a timeout. This will not
+introduce duplicates if the client initiates the submission with a uniquely
+chosen transaction ID. Since the server keeps a durable log of all transaction
+IDs, it will not accept an already completed transaction again. Alternatively,
+the server can generate a unique transaction ID and require the client to
+provide that ID during the final submission. Both approaches have the same
+effect, and we leave the implementation detail to the specific metastore at
+hand.
```

Comment:
@tobim this is the only content delta of my commit. The rest is wordsmithing.

Reply:
I think server-issued transaction ids are an interesting idea, but they seem pretty under specified right now. Naively, if the server crashes the state that supports the transaction id generation may get lost, and we're back to square one.

Reply:
But when the client commits a transaction, with either their own or server-generated ID, and doesn't get an ACK, the client would retry regardless with the same ID.

Reply:
The latest addition clears my concerns.